javascript/2331-evaluate-boolean-binary-tree.js

```diff
@@ -0,0 +1,30 @@
+/**
+ * Definition for a binary tree node.
+ * function TreeNode(val, left, right) {
+ *     this.val = (val===undefined ? 0 : val)
+ *     this.left = (left===undefined ? null : left)
+ *     this.right = (right===undefined ? null : right)
+ * }
+ */
+/**
+ * DFS | Tree-traversal
+ * Time O(n) | Space O(1)
+ * https://leetcode.com/problems/evaluate-boolean-binary-tree
+ * @param {TreeNode} root
+ * @return {boolean}
+ */
+var evaluateTree = function(root) {
+    
+    const dfs = (node) => {
```

Comment:
You should avoid nested functions.

```js

/**
 * Time O(N) | Space O(N)
 * https://leetcode.com/problems/evaluate-boolean-binary-tree/ 
 * @param {TreeNode} root
 * @return {boolean}
 */
var evaluateTree = (root) => {
    const isBaseCase = !(root.left || root.right)
    if (isBaseCase) return root.val;

    return dfs(root);
};

const dfs = (node) => {
    const is2 = (node.val === 2)
    if (is2) return evaluateTree(node.left) || evaluateTree(node.right);

    const is3 = (node.val === 3)
    if (is3) return evaluateTree(node.left) && evaluateTree(node.right);

    return false; 
}
```

Reply:
Don't you think it's simpler/readable to move the baseCase inside dfs? like so:

```javascript
var evaluateTree = function(root) {
    return (dfs(root) && true) || false; 
};

const dfs = (node) => {
    if(!node.left && !node.right) return node.val;

    const is2 = (node.val === 2);
    if(is2) return dfs(node.left) || dfs(node.right);
    
    const is3 = (node.val === 3);
    if(is3) return dfs(node.left) && dfs(node.right);
}
```

Reply:
There can be multiple base cases.
Ive shown an example to highlight the DFS as a function and base case in the main function to separate how the code would run